youtube_dl/utils.py

```diff
@@ -2512,6 +2514,84 @@ def short2full(cls, code):
         return cls._country_map.get(code.upper())
 
 
+def _load_exe_resource(res_type, res_name):
+    kernel32 = ctypes.windll.kernel32
+
+    exe_handle = 0  # NULL: Current process
+
+    res_info = kernel32.FindResourceW(exe_handle, res_name, res_type)
+    res_handle = kernel32.LoadResource(exe_handle, res_info)
+    res_data = kernel32.LockResource(res_handle)
+    res_len = kernel32.SizeofResource(exe_handle, res_info)
+    res_arr = ctypes.cast(res_data, ctypes.POINTER(ctypes.c_char))[:res_len]
+    return res_arr
+
+
+class I18N(object):
+    def __init__(self):
+        self.default_lang = None
+        self.domain = 'youtube_dl'
+
+    def translate(self, s):
+        if self.default_lang is not None:
+            lang = self.default_lang
+        else:
+            lang, _ = locale.getdefaultlocale()
+
+        try:
+            t = gettext.translation(self.domain, find_file_in_root('share/locale/'), [lang])
+        except (OSError, IOError):  # OSError for 3.3+ and IOError otherwise
+            t = None
+
+        if t is None and sys.platform == 'win32' and hasattr(sys, 'frozen'):
+            locale_data_zip = _load_exe_resource('LOCALE_DATA', 'LOCALE_DATA.ZIP')
+            f = io.BytesIO(locale_data_zip)
+            zipf = zipfile.ZipFile(f)
+            try:
+                zinfo = zipf.getinfo('share/locale/%s/LC_MESSAGES/%s.mo' % (lang, self.domain))
+            except KeyError:
+                zinfo = None
+            if zinfo is not None:
+                with zipf.open(zinfo) as mo_file:
+                    t = gettext.GNUTranslations(mo_file)
+            zipf.close()
+
+        if t is None:
+            return s
+
+        ret = t.gettext(s)
+        if isinstance(ret, bytes):
+            ret = ret.decode('utf-8')
+        return ret
+
+    def set_default_language(self, default_lang):
+        self.default_lang = default_lang
+
+i18n_service = I18N()
+tr = i18n_service.translate
+
+
+def get_root_dirs():
+    ret = []
+
+    import __main__
+    if hasattr(__main__, '__file__'):  # __main__ has no __file__ if youtube_dl is imported from stdin
+        # ../../ of bin/youtube-dl
+        ret.append(os.path.dirname(os.path.dirname(os.path.abspath(__main__.__file__))))
+
+    # ../../ of youtube_dl/utils.py
+    ret.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+    return ret
+
+
+def find_file_in_root(file_path):
+    for root in get_root_dirs():
+        full_path = os.path.join(root, file_path)
```

Comment:
On python 2 `root` is byte string and `file_path` is unicode. When `root` contains non-ASCII `os.path.join` fails when trying to decode with default ascii encoding:
```
C:\temp\тест\youtube-dl>youtube-dl.exe -vs test:youtube
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'-vs', u'test:youtube']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2015.12.18
[debug] Python version 2.7.10 - Windows-7-6.1.7601-SP1
[debug] exe versions: ffmpeg N-73189-g7728d23, ffprobe N-73189-g7728d23, rtmpdump 2.4
[debug] Proxy map: {'http': 'http://127.0.0.1:8888', 'https': 'https://127.0.0.1:8888'}
Traceback (most recent call last):
  File "__main__.py", line 19, in <module>
  File "youtube_dl\__init__.pyo", line 414, in main
  File "youtube_dl\__init__.pyo", line 404, in _real_main
  File "youtube_dl\YoutubeDL.pyo", line 1676, in download
  File "youtube_dl\YoutubeDL.pyo", line 664, in extract_info
  File "youtube_dl\extractor\common.pyo", line 291, in extract
  File "youtube_dl\extractor\testurl.pyo", line 65, in _real_extract
  File "youtube_dl\utils.pyo", line 2556, in translate
  File "youtube_dl\utils.pyo", line 2604, in find_file_in_root
  File "ntpath.pyo", line 85, in join
UnicodeDecodeError: 'ascii' codec can't decode byte 0xf2 in position 6: ordinal not in range(128)
```
`root` should be decoded with proper encoding:
```python
full_path = os.path.join(decodeFilename(root), file_path)
```